src/devices/src/legacy/serial.rs

```diff
@@ -276,34 +288,72 @@ impl BusDevice for Serial {
 }
 
 impl Subscriber for Serial {
-    /// Handle a read event (EPOLLIN) on the serial input fd.
-    fn process(&mut self, event: &EpollEvent, _: &mut EventManager) {
+    /// Handle events on the serial input fd.
+    fn process(&mut self, event: &EpollEvent, ev_mgr: &mut EventManager) {
         let source = event.fd();
         let event_set = event.event_set();
 
-        // TODO: also check for errors. Pending high level discussions on how we want
-        // to handle errors in devices.
-        let supported_events = EventSet::IN;
-        if !supported_events.contains(event_set) {
+        // We expect to be interested only in serial input.
+        let interest_list = self.interest_list();
+        if interest_list.len() != 1 {
+            return;
+        }
+
+        // Safe to unwrap. Checked before if interest list has one element.
+        let supported_event = interest_list.first().unwrap();
+
+        if !supported_event.event_set().contains(event_set) {
             warn!(
                 "Received unknown event: {:?} from source: {:?}",
                 event_set, source
             );
             return;
         }
 
-        if let Some(input) = self.input.as_mut() {
-            if input.as_raw_fd() == source {
-                let mut out = [0u8; 32];
-                match input.read(&mut out[..]) {
+        // Check if the event source is the serial input.
+        if supported_event.fd() != source {
+            return;
+        }
+
+        // Handle `EventSet::IN` to dispose inflight bytes.
+        if event_set.contains(EventSet::IN) {
+            match self.recv_bytes() {
+                Ok(_) => {}
+                Err(e) => warn!("error while reading stdin: {:?}", e),
+            }
+        }
+
+        // Once `EventSet::IN` was successfully handled, handle other events.
+        // Unregister the event source if peer fd was closed: `EventSet::HANG_UP`.
+        // http://man7.org/linux/man-pages/man2/epoll_ctl.2.html
+        if event_set.contains(EventSet::HANG_UP) {
+            // Read inflight bytes if any. Keep reading until we get all of them
+            // consumed.
+            loop {
+                match self.recv_bytes() {
                     Ok(count) => {
-                        self.raw_input(&out[..count])
-                            .unwrap_or_else(|e| warn!("Serial error on input: {}", e));
+                        if count == 0 {
+                            break;
+                        }
                     }
                     Err(e) => {
                         warn!("error while reading stdin: {:?}", e);
+                        break;
                     }
-                }
+                };
+            }
```

Comment:
Do we need to consume pending bytes? I don't see an obvious reason to do so if we unregister the fd anyway in the next step.
(less code is better)